streams/src/main/java/org/apache/kafka/streams/state/RocksDBConfigSetter.java

```diff
@@ -37,4 +37,19 @@
      * @param configs       the configuration supplied to {@link org.apache.kafka.streams.StreamsConfig}
      */
     void setConfig(final String storeName, final Options options, final Map<String, Object> configs);
+
+    /**
+     * This method will be called when the store is closed and should be used to close any user-constructed objects
+     * that inherit from RocksObject. Any such object created with new in setConfig should have close called on it
+     * here to avoid leaking off-heap memory.
+     * Objects to be closed can be saved by the user or retrieved back from options using its getter methods.
+     *
+     * Example objects needing to be closed include Filter and Cache
+     *
+     * @param storeName     the name of the store being configured
+     * @param options       the Rocks DB options
+\    */
+    default void close(final String storeName, final Options options) {
```

Comment:
That's a tricky question... The problem is that there's no way to signal ahead of time that you *should* override the method to avoid breakage when we remove `default` (unlike when you deprecate a method, which indicates you should make a change ahead of the breakage).

So, whenever we do make the change, it will be a breaking one.

If we think it's very likely that implementers would have some objects that need to be closed, and thus are leaking memory today, maybe we should actually just not default it at all, so they fix their memory leaks asap.

Conversely, if we think it's unlikely that the average implementer is going to leak memory, there's no reason we shouldn't just leave it defaulted forever.

Reply:
We need the default implementation in any case until the next major release not to break existing code. Afterwards, I would be in favour of removing the default implementation, because then there is a bigger chance that existing memory leakages are found and future leakages are avoided.

Reply:
Hm. On the one hand I'd guess not that many users **in total** are leaking memory here, but suspect that those who are advanced enough to be implementing ConfigSetter may likely be... 

I'd agree we should remove the default in the next major release though, especially once we upgrade Rocks as some of the API changes since our current version effectively force creating more/new RocksObjects when setting options.

Reply:
Might be worth to add a WARN log in the default implementation, telling users, that the default will be removed in 3.0.0 and they should overwrite it?

Reply:
Good call